tree/30_generic_methods/logger_rudder.cf

```diff
@@ -33,14 +33,14 @@ bundle agent logger_rudder(message, class_prefix)
       "expected_reports_temp"   string => "${expected_reports_source}.tmp";
       "expected_reports_file"   string => "${expected_reports_source}.res";
 
-    logger_rudder_temp_resfile_repaired::
+    (!final_resfile_exists.logger_rudder_temp_resfile_reached|logger_rudder_temp_resfile_repaired)::
       # 2/ If the temporary file has been updated, read it to get all values in an array, and canonify the first entry
       "dim" int => readstringarrayidx("reports", "${expected_reports_temp}", "\s*#[^\n]*", ";;", 9999, 999999);
       "keys" slist => getindices("reports");
 
       "reports[${keys}][canon]" string => canonify("${reports[${keys}][1]}");
 
-    (logger_rudder_final_resfile_repaired|logger_rudder_temp_resfile_kept)::
+    (logger_rudder_final_resfile_repaired|logger_rudder_temp_resfile_kept.!logger_rudder_temp_resfile_repaired.final_resfile_exists)::
       "c_class_prefix" string => canonify("${class_prefix}");
```

Comment:
One of the challenges here is not entering this "block" too early - we need to only get into on the third pass (pass 1: check if the temporary file needs to be updated, pass 2: read keys from temporary file + write the final expected reports). Therefore we must make sure other promises have been evaluated first (see below).

Put simply, we need to enter this block if and only if we know we've successfully obtained the final expected reports file. This can happen for one of two reasons: **1)** it needing fixing and was fixed (easy: ```logger_rudder_final_resfile_repaired```) or **2)** it didn't need fixing but it is OK (as in we've already checked it).

The second reason is harder to get a class for. The ```logger_rudder_final_resfile_kept``` is not enough, because it may not be defined, since it is a conditional block (if ```final_resfile_exists``` and ```!logger_rudder_temp_resfile_repaired```). In effect, you could therefore say either we have the ```logger_rudder_final_resfile_kept``` class or we have the condition that means that promise was entirely skipped: ```logger_rudder_final_resfile_kept|(final_resfile_exists.!logger_rudder_temp_resfile_repaired)```.

Your example (in case when resfile was correct but tmp file was not) would lead to ```logger_rudder_temp_resfile_repaired```, and therefore the ```logger_rudder_final_resfile_kept``` class would be set, because that promise would be reached.

Therefore we should be able to use:

```
logger_rudder_final_resfile_repaired|logger_rudder_final_resfile_kept|(final_resfile_exists.!logger_rudder_temp_resfile_repaired)
```

BUT this last condition (```(final_resfile_exists.!logger_rudder_temp_resfile_repaired)```) can cause this block to be entered on pass 2, too early. I suggest we simply add in a pass1/pass2/pass3 set of classes and use:

```
logger_rudder_final_resfile_repaired|logger_rudder_final_resfile_kept|(pass3.final_resfile_exists.!logger_rudder_temp_resfile_repaired)
```

WDYT?